chain/default_syncer_test.go

```diff
@@ -858,48 +859,58 @@ func TestTipSetWeightDeep(t *testing.T) {
-	measuredWeight, err := wFun(chain.Head())
+	assertHead(assert, chainStore, tsShared)
+	measuredWeight, err := wFun(chainStore.Head())
 	require.NoError(err)
 	expectedWeight := startingWeight + uint64(22000)
 	assert.Equal(expectedWeight, measuredWeight)
 
 	// fork 1 is heavier than the old head.
-	f1b2a := RequireMkFakeChildCore(require,
-		FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f1b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, MinerAddr: info.Miners[1].Address},
+	f1b2a := chain.RequireMkFakeChildCore(require,
+		chain.FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f1b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, MinerAddr: info.Miners[1].Address},
 		wFun)
-	f1b2a.Proof, f1b2a.Ticket, err = MakeProofAndWinningTicket(info.Miners[1].Address, info.Miners[1].Power, 1000)
+	f1b2a.Proof, f1b2a.Ticket, err = chain.MakeProofAndWinningTicket(info.Miners[1].Address, info.Miners[1].Power, 1000)
 	require.NoError(err)
 
-	f1b2b := RequireMkFakeChildCore(require,
-		FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f1b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, Nonce: uint64(1), MinerAddr: info.Miners[2].Address},
+	f1b2b := chain.RequireMkFakeChildCore(require,
+		chain.FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f1b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, Nonce: uint64(1), MinerAddr: info.Miners[2].Address},
 		wFun)
-	f1b2b.Proof, f1b2b.Ticket, err = MakeProofAndWinningTicket(info.Miners[2].Address, info.Miners[2].Power, 1000)
+	f1b2b.Proof, f1b2b.Ticket, err = chain.MakeProofAndWinningTicket(info.Miners[2].Address, info.Miners[2].Power, 1000)
 	require.NoError(err)
 
 	f1 := testhelpers.RequireNewTipSet(require, f1b2a, f1b2b)
 	f1Cids := requirePutBlocks(require, cst, f1.ToSlice()...)
 	err = syncer.HandleNewBlocks(ctx, f1Cids)
 	require.NoError(err)
-	assertHead(assert, chain, f1)
-	measuredWeight, err = wFun(chain.Head())
+	assertHead(assert, chainStore, f1)
+	measuredWeight, err = wFun(chainStore.Head())
 	require.NoError(err)
 	expectedWeight = startingWeight + uint64(33000)
 	assert.Equal(expectedWeight, measuredWeight)
 
 	// fork 2 has heavier weight because of addr3's power even though there
 	// are fewer blocks in the tipset than fork 1.
-	f2b2 := RequireMkFakeChildCore(require,
-		FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f2b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, MinerAddr: info.Miners[3].Address},
+	f2b2 := chain.RequireMkFakeChildCore(require,
+		chain.FakeChildParams{Parent: testhelpers.RequireNewTipSet(require, f2b1), GenesisCid: calcGenBlk.Cid(), StateRoot: bootstrapStateRoot, MinerAddr: info.Miners[3].Address},
 		wFun)
+	// This should fix https://github.com/filecoin-project/go-filecoin/issues/1828
+	f2b2.Proof, f2b2.Ticket, err = chain.MakeProofAndWinningTicket(info.Miners[3].Address, info.Miners[3].Power, 1000)
```

Comment:
can you splain what the issue was and the fix? like, how did this pass and what caused the nondeterminism?

Reply:
My best guess is it was because that block's miner power was 980 compared to 1000 total Power, so most of the time it was a winning ticket because of the extremely large "stake".